### DISCUSSION TITLE
Adding a way to enable code spelling check

### DISCUSSION THREAD

**Discussion by alexsmartens** [nvim-treesitter/nvim-treesitter#6877] [Ideas]
## Problem statement
`Treesitter` only spell checks comments (e.g., [ruby/highlight.scm](https://github.com/nvim-treesitter/nvim-treesitter/blob/a8c082b4f5b3b70bde2ffb08cc112dfd7e515fb3/queries/ruby/highlights.scm#L207)). At the same time many people would like to have their code spell checked, including checking method names, variables and what not. For some reason there is no straightforward way to do that.

### Impact
For reference, [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) plugin that adds code spelling check to VSCode was installed **11M** times. Perhaps some `Treesitter` users would appreciate this feature.

## Idea

Add a way to globally enable code spelling check in the configuration, perhaps it might look something like this:
```
...
highlight = {
  enable = true,
  spell = {
    check_code = true,
  },
},
...
```
The above configuration would set [@spell ](https://github.com/nvim-treesitter/nvim-treesitter/blob/7ff0b8d3d21b15eb7231fcbd2acfded8987f5efd/CONTRIBUTING.md?plain=1#L260) for every identifier under the hood. If needed I would be happy to raise a PR to make this behavior work, I would just need someone to point me to a similar PR for changing the configuration.

## Workaround

https://github.com/alexsmartens/dotfiles/commit/b0486cd3a6b81176a94c730cf7ff2f3f41c482df - a commit to my dot files. 

Enable `nvim-treesitter` [spell checking](https://github.com/nvim-treesitter/nvim-treesitter/blob/7ff0b8d3d21b15eb7231fcbd2acfded8987f5efd/CONTRIBUTING.md?plain=1#L260) for all nodes: 
```scm
; .config/nvim/queries/ruby/highlights.scm

; Spell check every node
 ((_) @spell)

; Copy the default config's content below or otherwise all the default configurations would be lost
; https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/ruby/highlights.scm
```

Done!

### Note
`highlights.scm` configuration is language specific so you'd have to change it for every language you work with.

**Comment by clason** [comment]
This is out of scope for this repo. The default experience people want is more targeted. Feel free to open an enhancement request at vim for bypassing all syntax restrictions for spelling.

**Comment by alexsmartens** [comment]
**11M** downloads of the [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) VSCode plugin speaks for itself. Perhaps this idea is worth considering

**Reply by clason** [reply]
I could make a tasteless joke about this, but will leave it at pointing out you're comparing apples and oranges: is the Code Spell Checker VSCode plugin built on treesitter queries?

**Reply by alexsmartens** [reply]
That's obvious, dude. Not sure why you keep pushing back with no good reason.

Perhaps you'd want to explain your point without your funny jokes. Otherwise feel free to ignore

**Reply by clason** [reply]
If it's obvious, why are you raising this issue here where it's not actionable? I really don't understand your point.

And my reason is very simple: I'm explaining where the best place to ask for your feature is -- if you did not want a response, why did you ask in the first place? I'm sorry that this is not the answer you hoped for, but things are what they are.

**Reply by alexsmartens** [reply]
I dug depper. Nvim configuration is irrelevant. [@spell](https://github.com/nvim-treesitter/nvim-treesitter/blob/7ff0b8d3d21b15eb7231fcbd2acfded8987f5efd/CONTRIBUTING.md?plain=1#L260) is a Treesitter's concept. It is set in `highlights.scm` (e.g., [ruby/highlight.scm](https://github.com/nvim-treesitter/nvim-treesitter/blob/a8c082b4f5b3b70bde2ffb08cc112dfd7e515fb3/queries/ruby/highlights.scm#L207)). Given the config is language specific it'd be pretty painful to manage locally, especially for folks willing to keep up with most recent highlights. I've also updated the description with my findings.

Doesn't my proposal seem relevant now?

**Reply by clason** [reply]
No. Spelling _is_ a Neovim concept; the `@spell` capture just restricts that feature to certain nodes (similar to how legacy syntax files work). What you want is to _not_ restrict spelling in this way, so this option(!) needs to be implemented where spelling happens. Since the spelling code is taken from Vim, and the feature applies to legacy syntax in exactly the same way, that's where the feature request should be made.

Similarly query management is handled by Neovim, _not_ this plugin (which only contains the language-specific queries themselves). So "global" extensions need to be implemented there.

Again (and please trust the maintainer on this): Your feature **CANNOT** be implemented in this plugin.

**Reply by kopiczko** [reply]
> Again (and please trust the maintainer on this): Your feature **CANNOT** be implemented in this plugin.

That's true, but do you think it would be somehow possible to have a setting to add `@spell` to other groups, e.g. `@variable`? I guess that would require some dynamic `*.scm` building.

**Reply by clason** [reply]
Not in this plugin. Conditional queries _can_ be implemented in Neovim (I had a PR up), but this would not work in other editors so we decided not to do it to avoid even further fragmentation.

**Reply by theHamsta** [reply]
Honestly, I think not having any `@spell` or queries at all would be for me a better experience (had a long time a custom Neovim with this code path disabled). Since this would not require any queries at all, there should be a way to disable selected spell checking via queries. I believe the "code-only" spell checking via syntax files is a feature that you currently can opt-in, but also disable if you don't like. Having tree-sitter working similar like syntax could be considered for Neovim when somebody creates a PR

**Reply by clason** [reply]
>  Since this would not require any queries at all, there should be a way to disable selected spell checking via queries.

You can always `extend` queries with `@nospell`.

And just to be clear: any change to spell-checking in Neovim must go through `:set spelloptions`; tree-sitter is just an implementation detail. So

> having tree-sitter working similar like syntax could be considered for Neovim when somebody creates a PR

is in fact how it works already.


The only relevant change I can see upstream is support for "generic queries" (or allowing to extend queries via `query.set()`) so you can override easily for all languages.

**Reply by theHamsta** [reply]
Before, spell checking via syntax was configurable to ignore syntax
```
SPELL CHECKING						*:syn-spell*

:sy[ntax] spell toplevel
:sy[ntax] spell notoplevel
:sy[ntax] spell default
	This defines where spell checking is to be done for text that is not
	in a syntax item:

	toplevel:	Text is spell checked.
	notoplevel:	Text is not spell checked.
	default:	When there is a @Spell cluster no spell checking.

	For text in syntax items use the @Spell and @NoSpell clusters
	|spell-syntax|.  When there is no @Spell and no @NoSpell cluster then
	spell checking is done for "default" and "toplevel".

	To activate spell checking the 'spell' option must be set.
```

**Reply by clason** [reply]
Sorry, that doesn't do what you say: It will not ignore `@spell` and `@nospell`; it only affects text that is not contained in a syntax region (which never applies for tree-sitter due to the tree).

**Reply by alexsmartens** [reply]
> The only relevant change I can see upstream is support for "generic queries" (or allowing to extend queries via query.set()) so you can override easily for all languages.

@clason that would be very appreciated 🙌
